File: 3-longest-substring-without-repeating-characters/3-longest-substring-without-repeating-characters.py
from collections import Counter
class Solution(object):
    def lengthOfLongestSubstring(self, s): 
            if len(s) == 0:
                return 0
            Holds = {}
            maxLength = start = 0
            for i in range(len(s)):

                if s[i] in Holds and start <= Holds[s[i]]:
                    # print(Holds[s[i]],end= '   ')
                    start = Holds[s[i]] + 1
                    # print(start)
                else:
                    maxLength = max(maxLength, i - start + 1)
                Holds[s[i]] = i
            print(Holds)
            return maxLength

File: 3-longest-substring-without-repeating-characters/3-longest-substring-without-repeating-characters.java
class Solution {
    public int lengthOfLongestSubstring(String s) {
     if(s.length()<=1) return s.length();
    int L=0,R=0,maxLen=0,N=s.length();
    char[] arr=s.toCharArray();
    Map<Character,Integer> hm= new HashMap<>();
    while (R<N){
      hm.merge(arr[R],1, (oldValue, newValue) -> oldValue + 1);

      while (hm.size()!=R-L+1){
        hm.merge(arr[L],hm.get(arr[L]),(oldValue, newValue) -> oldValue -1);
        if(hm.get(arr[L])==0)hm.remove(arr[L]);
        L+=1;
      }
      maxLen=Math.max(maxLen,R-L+1);
//      maxLen=Math.max(maxLen,hm.size());
      R++;
    }
   
        return  maxLen;
  }
}